### PR TITLE
[Merged by Bors] - 2885 p2p collect metrics

### LINKED ISSUE #2885: p2p: collect metrics
libp2p integration was merged without metrics, we should collect next metrics:

- total number of peers
- number of peers per each gossip protocol
- total sent/received traffic
- traffic for each gossip protocol (protocol is a label, sent/received is a label)
- number of messages for each gossip protocol
- number of streams per each protocol
- number of connections
- histogram for server request durations for each protocol

Libp2p provides tracing infrastructure that allows to plug custom metrics collectors. 
For gossipsub see pubsub.RawTracer interface.
For bandwidth  see metrics.BandwidthReporter interface, that is enabled with libp2p.BandwidthReporter(). 
For streams and connections see interface that is expected by host.Network().Notify method.

### PR DESCRIPTION
**Pull request by abergasov**
## Motivation

Closes #3193

## Changes
Collect metrics for prometheus from p2p protocol

## DevOps Notes

- [x] This PR does not require configuration changes (e.g., environment variables, GitHub secrets, VM resources)
- [x] This PR does not affect public APIs
- [x] This PR does not rely on a new version of external services (PoET, elasticsearch, etc.)
- [x] This PR does not make changes to log messages (which monitoring infrastructure may rely on)

### CODE REVIEW

**Comment by dshulyak** [comment]
we should use prometheus to track metrics, please check examples in other modules

File: p2p/metrics/gossipsub.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+var (
+	totalPeers       = metrics.NewGauge("total_peers", subsystem, "Total number of peers", nil)
+	peersPerProtocol = metrics.NewGauge("peers_per_protocol", subsystem, "Number of peers per protocol", []string{"protocol"})
+)
+
+// GossipCollector pubsub.RawTracer implementation
+// total number of peers
+// number of peers per each gossip protocol.
+type GossipCollector struct {
+	peers struct {
```
Comment (dshulyak) [inline p2p/metrics/gossipsub.go:22]: maybe it can be removed?

Comment (abergasov) [reply p2p/metrics/gossipsub.go:22]: no, it need to track when peer is removing. Interface does not provide protocol of removed peer
so we need track this info on our side

**Review by dshulyak — APPROVED** [review]

File: p2p/metrics/bandwidth.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+// BandwidthCollector implement metrics.Reporter
+// that keeps track of the number of messages sent and received per protocol.
+type BandwidthCollector struct{}
+
+// NewBandwidthCollector creates a new BandwidthCollector.
+func NewBandwidthCollector() *BandwidthCollector {
+	return &BandwidthCollector{}
+}
+
+// LogSentMessageStream logs the message sent by the peer.
```
Comment (countvonzero) [inline p2p/metrics/bandwidth.go]: by the node or by the peer? same for the rest of the comments in this file.

File: p2p/metrics/bandwidth.go [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+	messagesPerProtocol.WithLabelValues(string(proto), incoming).Inc()
+}
+
+// LogSentMessage  logs the message sent by the peer.
+func (b *BandwidthCollector) LogSentMessage(int64) {}
+
+// LogRecvMessage logs the message received by the peer.
+func (b *BandwidthCollector) LogRecvMessage(int64) {}
+
+// GetBandwidthForPeer mock returns the bandwidth for a given peer.
```
Comment (countvonzero) [inline p2p/metrics/bandwidth.go:64]: mock? i am confused. are these methods used somewhere?

File: p2p/metrics/gossipsub.go [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+	totalPeers.WithLabelValues().Dec()
+}
+
+// Join is invoked when a new topic is joined.
+func (g *GossipCollector) Join(string) {}
+
+// Leave is invoked when a topic is abandoned.
+func (g *GossipCollector) Leave(string) {}
+
+// Graft is invoked when a new peer is grafted on the mesh (gossipsub).
```
Comment (countvonzero) [inline p2p/metrics/gossipsub.go:67]: apologies for my ignorance. what is the difference btwn leave/graft/prune?

File: p2p/metrics/bandwidth.go (hunk 3 above)
Comment (abergasov) [reply p2p/metrics/bandwidth.go:64]: yep, we not use this methods. 
It implement interface just for several methods, which allow collect metrics. for other methods - mock return

File: p2p/metrics/bandwidth.go (hunk 2 above)
Comment (abergasov) [reply p2p/metrics/bandwidth.go]: rewrite comments for better understanding
in this case we register message that our node send to some peer

File: p2p/metrics/gossipsub.go (hunk 4 above)
Comment (abergasov) [reply p2p/metrics/gossipsub.go:67]: connection between peers can be `full-messages` or `meta-only` at one time 
so if peers want upgrade from `meta` to `full` - they run `Graft` procedure
if they wanna downgrade from `full` to `meta` - they run `Prune`

Leave - when peer leave topic - he send message to all connected peers that `i'm not subscribed to this topic anymore`

**Review by dshulyak — APPROVED** [review]

**Review by countvonzero — APPROVED** [review]
thanks for the answers.

**Comment by abergasov** [comment]
bors try

**Comment by abergasov** [comment]
bors merge

**Pull request closed without merging**